File: airbyte-cdk/java/airbyte-cdk/db-sources/src/main/java/io/airbyte/cdk/integrations/debezium/internals/RelationalDbDebeziumEventConverter.java
/*
 * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.cdk.integrations.debezium.internals;

import java.time.Instant;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;

import io.airbyte.cdk.integrations.debezium.CdcMetadataInjector;
import io.airbyte.protocol.models.v0.AirbyteMessage;

public class RelationalDbDebeziumEventConverter implements DebeziumEventConverter {

  private final CdcMetadataInjector cdcMetadataInjector;
  private final Instant emittedAt;
  private final Map<String, List<String>> jsonbColumnsPerTable;
  private final ObjectMapper objectMapper = new ObjectMapper();

  public RelationalDbDebeziumEventConverter(CdcMetadataInjector cdcMetadataInjector, Instant emittedAt) {
    this.cdcMetadataInjector = cdcMetadataInjector;
    this.emittedAt = emittedAt;
    this.jsonbColumnsPerTable = new HashMap<>();
  }

  public RelationalDbDebeziumEventConverter(CdcMetadataInjector cdcMetadataInjector, Instant emittedAt,
      Map<String, List<String>> jsonbColumnsPerTable) {
    this.cdcMetadataInjector = cdcMetadataInjector;
    this.emittedAt = emittedAt;
    this.jsonbColumnsPerTable = jsonbColumnsPerTable;
  }

  @Override
  public AirbyteMessage toAirbyteMessage(ChangeEventWithMetadata event) {
    final JsonNode debeziumEvent = event.eventValueAsJson();
    final JsonNode before = debeziumEvent.get(DebeziumEventConverter.BEFORE_EVENT);
    final JsonNode after = debeziumEvent.get(DebeziumEventConverter.AFTER_EVENT);
    final JsonNode source = debeziumEvent.get(DebeziumEventConverter.SOURCE_EVENT);

    final ObjectNode baseNode = (ObjectNode) (after.isNull() ? before : after);
    convertJsonb(baseNode, source);
    final JsonNode data = DebeziumEventConverter.addCdcMetadata(baseNode, source, cdcMetadataInjector, after.isNull());
    return DebeziumEventConverter.buildAirbyteMessage(source, cdcMetadataInjector, emittedAt, data);
  }

  private void convertJsonb(final ObjectNode data, final JsonNode source) {
    final String tableName = source.get("table").asText();
    List<String> jsonbColumns = this.jsonbColumnsPerTable.get(tableName);
    if (jsonbColumns != null) {
      jsonbColumns.forEach(c -> {
        final String stringifiedJsonb = data.get(c).asText();
        try {
          data.set(c, this.objectMapper.readTree(stringifiedJsonb));
        } catch (JsonProcessingException e) {
          // should this throw or silently swallow?
          throw new RuntimeException("Could not parse 'jsonb' value:" + e);
        }
      });
    }
  }

}

File: airbyte-cdk/java/airbyte-cdk/db-sources/src/main/kotlin/io/airbyte/cdk/integrations/debezium/internals/RelationalDbDebeziumEventConverter.kt
/*
 * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
 */
package io.airbyte.cdk.integrations.debezium.internals

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.node.ObjectNode
import com.fasterxml.jackson.databind.ObjectMapper;
import io.airbyte.cdk.integrations.debezium.CdcMetadataInjector
import io.airbyte.protocol.models.v0.AirbyteMessage
import java.time.Instant


class RelationalDbDebeziumEventConverter(
    private val cdcMetadataInjector: CdcMetadataInjector<*>,
    private val emittedAt: Instant,
    private val jsonbColumnsPerTable: Map<String, List<String>> = HashMap<String, List<String>>()
) : DebeziumEventConverter {
    private val objectMapper: ObjectMapper = ObjectMapper();
    override fun toAirbyteMessage(event: ChangeEventWithMetadata): AirbyteMessage {
        val debeziumEvent = event.eventValueAsJson()
        val before: JsonNode = debeziumEvent!!.get(DebeziumEventConverter.Companion.BEFORE_EVENT)
        val after: JsonNode = debeziumEvent.get(DebeziumEventConverter.Companion.AFTER_EVENT)
        val source: JsonNode = debeziumEvent.get(DebeziumEventConverter.Companion.SOURCE_EVENT)

        val baseNode = (if (after.isNull) before else after) as ObjectNode
        convertJsonb(baseNode, source, jsonbColumnsPerTable);
        val data: JsonNode =
            DebeziumEventConverter.Companion.addCdcMetadata(
                baseNode,
                source,
                cdcMetadataInjector,
                after.isNull
            )
        return DebeziumEventConverter.Companion.buildAirbyteMessage(
            source,
            cdcMetadataInjector,
            emittedAt,
            data
        )
    }

    private fun convertJsonb(data: ObjectNode, source: JsonNode, jsonbColumnsPerTable: Map<String, List<String>>) {
    val tableName: String = source.get("table").asText();
    val jsonbColumns: List<String>? = jsonbColumnsPerTable.get(tableName);
    if (jsonbColumns != null) {
        for (c in jsonbColumns)  {
            val stringifiedJsonb: String = data.get(c).asText();
            try {
            data.set(c, objectMapper.readTree(stringifiedJsonb));
            } catch (e: JsonProcessingException) {
            // should this throw or silently swallow?
            throw RuntimeException("Could not parse 'jsonb' value:" + e);
            }
        }
    }
  }
}
